### DISCUSSION TITLE
"qtile-extras" - stuff I've been playing around with

### DISCUSSION THREAD

**Discussion by elParaguayo** [qtile/qtile#2805] [Show and tell]
I've posted a few times about some of the items that I use for my own build that probably won't get merged into the main repo.

I've finally got round to grouping these all into a separate [repo](https://github.com/elParaguayo/qtile-extras) and got some decent [documentation](https://qtile-extras.readthedocs.io/) in place so I thought it would be good to share here.

The following widgets are available:

<table>
<tr>
<td width="20%">ALSAWidget</td><td width="80%">
<img src=https://user-images.githubusercontent.com/946265/133936315-f869c6d5-dfcc-4130-9517-60159d6e9403.gif />
</td>
</tr>
<tr>
<td>BrightnessControl</td><td>
<img src=https://user-images.githubusercontent.com/946265/133936337-e590cf0a-830f-4e9d-ae90-8c761c2e6c4c.gif />
</td>
</tr>
<tr>
<td>LiveFootballScores</td><td>
<img src=https://user-images.githubusercontent.com/946265/133936351-09926b39-46bc-4472-bd63-e7bd6a6f86e6.gif />
</td>
</tr>
<tr>
<td>ScriptExit</td><td>Like QuickExit but runs a script before exiting.</td>
</tr>
<tr>
<td>StravaWidget</td><td>
<img src=https://user-images.githubusercontent.com/946265/133937669-6a324ae5-ae62-417b-b69d-933373b821e3.png />
</td>
</tr>
<tr>
<td>TVHWidget</td><td>
<img src=https://user-images.githubusercontent.com/946265/133936372-41ff30d6-b16b-48be-a0bb-eef80fbcc563.gif />
</td>
</tr>
<tr>
<td>UPowerWidget</td><td>
<img src=https://user-images.githubusercontent.com/946265/133936414-987c1f11-4648-48f6-9159-d561a02e7e4c.gif />
</td>
</tr>
<tr>
<td>WiFiIcon</td><td>
<img src=https://user-images.githubusercontent.com/946265/133936392-913b738d-f3b3-4e83-b468-d6b9f24c5e68.png />
</td>
</tr>
<tr>
<td>WordClock</td><td>
<img src=https://user-images.githubusercontent.com/946265/134418601-c275fe3f-5669-4b7e-924a-6904792e873c.png />

</td>
</tr>
</table>

I've also put my popup toolkit in the repo. It's a work in progress but it can be made to make launchers and extend widgets.

For example, here's a power options menu I made:
![20210919_160524](https://user-images.githubusercontent.com/946265/133936467-dfde3d82-2d4d-4752-96fd-7d36762cc9cb.png)

And, before anyone asks, yes, the menu can be navigated via the keyboard!

**Comment by elParaguayo** [comment]
New widget added: WordClock

**Reply by m-col** [reply]
Looks awesome - how does the widget relate to the background?

**Reply by elParaguayo** [reply]
Ha. It really doesn't need to be a widget at all. It's zero width and just uses some of the widget's features (access to bar, timers etc.) for convenience. It also gives the user a more familiar way to customise the clock (colours, font etc).

The widget checks the time and, when there's an update, draws the new clock and then uses `screen.paint` to update the background.

**Reply by elParaguayo** [reply]
Weird issue with this one: if I don't include this line: https://github.com/elParaguayo/qtile-extras/blob/main/qtile_extras/widget/wordclock.py#L68 then the default `fontsize` is not respected. No idea what I'm missing...

**Reply by m-col** [reply]
> The widget checks the time and, when there's an update, draws the new clock and then uses screen.paint to update the background.

Lol nice, clever hack.

> the default fontsize is not respected

Yeah, strange. Does the value need accessing or setting? The += obviously does both but maybe knowing which one is needed will track down the logic .

**Reply by elParaguayo** [reply]
It just needs accessing. Another way to do it was with `logger.warning(f"{self.fontsize=}")`. Confused the hell out of me as I was trying to debug it and adding that line made it work!

**Comment by m-col** [comment]
The ALSA and brightness widgets look to have a shared meter for showing percentage. Is this a simple implementation? There a bunch of widgets that have a similar model (i.e. reflecting a percentage). Perhaps it'd be nice to add as a mixin or something?

**Reply by elParaguayo** [reply]
Yeah - that's on my list...

**Comment by m-col** [comment]
That power options menu is beautiful btw, when I next rejig my config I'm definitely going to use it!

**Reply by elParaguayo** [reply]
Yeah. I really like it and it's a good example of what can be done with the toolkit.

There's a small border on X11 which isn't there on Wayland which I need to look at too...
